python/mxnet/model.py

```diff
@@ -104,15 +105,18 @@ def _initialize_kvstore(kvstore, param_arrays, arg_params, param_names, update_o
 
 def _update_params_on_kvstore(param_arrays, grad_arrays, kvstore, param_names):
     """Perform update of param_arrays from grad_arrays on kvstore."""
-    for index, pair in enumerate(zip(param_arrays, grad_arrays)):
-        arg_list, grad_list = pair
-        if grad_list[0] is None:
-            continue
-        name = param_names[index]
+    size = len(grad_arrays)
+    start = 0
+    # Use aggregation by default only with NCCL
+    default_batch = 16 if 'nccl' in kvstore.type else 1
+    batch = int(os.getenv('MXNET_UPDATE_AGGREGATION_SIZE', default_batch))
+    while start < size:
+        end = start + batch if start + batch < size else size
         # push gradient, priority is negative index
-        kvstore.push(name, grad_list, priority=-index)
+        kvstore.push(param_names[start:end], grad_arrays[start:end], priority=-start)
```

Comment:
what's the purpose of this? Why should it be done in frontend?

Reply:
This enables aggregation of the reductions (in NCCL 2.1 it is not a real aggregation, since each reduction is being done in its own launch and they benefit from lack of synchronization between reductions in a group, but NCCL 2.2 will introduce real aggregation support). 
It needs to be done in the frontend, because kvstore itself does not have any information on which gradients should be aggregated, in which order and how much time is it supposed to wait before real launch - mxnet's dependency engine does not really allow for that. That's why we need to provide data about all of the reductions in a group to benefit from aggregation.

Reply:
Is it valid to push multiple keys in one call? If its valid why not push all keys in one call and decide aggregation size in backend?

Reply:
Also the `if grad_list[0] is None:` logic is removed.

Reply:
It is valid to push multiple keys in 1 call. But you still want to set priority to those reductions in frontend - kvstore should not assume the order in which you push gradients to it, so aggregating everything would get rid of priority altogether.
Regarding  the missing logic - good point, I forgot about it - will fix.